wazevo(docs): optimizing compiler (#2065)
tetratelabs/wazero: site/content/docs/how_the_optimizing_compiler_works/backend.md

that are **clobbered** by the body of the function. A register is clobbered
if its value is overwritten by the function, and it is not saved by the
callee. This information is used in the finalization phase to determine which
registers need to be spilled in the prologue. This is not strictly related
to register allocation in a textbook meaning, but it is a necessary step
for the finalization phase.

### Liveness Analysis

Intuitively, a variable or name binding can be considered _live_ at a certain
point in a program, if its value will be used in the future.

For instance:

```
1| int f(int x) {
2|   int y = 2 + x;
3|   int z = x + y;
4|   return z;
5| }
```

Variable `x` and `y` are both live at line 4, because they are used in the
expression `x + y` on line 3; variable `z` is live at line 4, because it is
used in the return statement.  However, variables `x` and `y` can be considered
_not_ live at line 4 because they are not used anywhere after line 3.

Statically, _liveness_ can be approximated by following paths backwards on the
control-flow graph, connecting the uses of a given variable to its definitions
(or its *unique* definition, assuming SSA form).

> In practice, while liveness is a property of each name binding at any point in
> the program, it is enough to keep track of liveness at the boundaries of basic
> blocks:
> 
> - the _live-in_ set for a given basic block is the set of all bindings that are
>   live at the entry of that block.
> - the _live-out_ set for a given basic block is the set of all bindings that
>   are live at the exit of that block. A binding is live at the exit of a block
> if it is live at the entry of a successor.
> 
> Because the CFG is a connected graph, it is enough to keep track of either
> live-in or live-out sets, and then propagate the liveness information backwards
> or forwards, respectively. In our case, we keep track of live-ins.

Author: would this work?

```suggestion
Because the CFG is a connected graph, it is enough to keep track of either
live-in or live-out sets, and then propagate the liveness information backward
or forward, respectively. In our case, we keep track of  live-in sets per block;
live-outs are derived from live-ins of the successor blocks when a block is allocated.
```

> or forwards, respectively. In our case, we keep track of live-ins.

Reviewer: works for me

### Allocation

We implemented a variant of the linear scan register allocation algorithm
described in [the Go compiler's allocator][go-regalloc].

Each basic block is allocated registers in a linear scan order, and the
allocation state is propagated from a given basic block to its successors.
Then, each block continues allocation from that initial state.

#### Merge States

Special care has to be taken when a block has multiple predecessors. We call
this *fixing merge states*: for instance, consider the following:

```goat { width="30%" }
 .---.     .---.
| BB0 |   | BB1 |
 '-+-'     '-+-'
   +----+----+
        |
        v
      .---.
     | BB1 |
      '---'
```

if the live-out set of a given block `BB0` is different from the live-out set
of a given block `BB1` and both are predecessors of a block `BB2`, then we need
to adjust `BB0` and `BB1` to ensure consistency with `BB2`. In practice, we
ensure that the registers that `BB2` expects to be live-in are live-out in

> `BB0` and `BB1`.

Reviewer: good

generic struct that comes already with an implementation of all or most of the
required methods.  For instance,`regalloc.Function`is implemented by
`backend.RegAllocFunction[*arm64.instruction, *arm64.machine]`.

`backend/isa/<arch>/abi.go` (where `<arch>` is either `arm64` or `amd64`)
contains the instantiation of the `regalloc.RegisterInfo` struct, which
declares, among others
- the set of registers that are available for allocation, excluding, for
  instance, those that might be reserved by the runtime or the OS
(`AllocatableRegisters`)
- the registers that might be saved by the callee to the stack
  (`CalleeSavedRegisters`)

### Debug Flags

- `wazevoapi.RegAllocLoggingEnabled` logs detailed logging of the register
  allocation procedure.
- `wazevoapi.PrintRegisterAllocated` prints the basic blocks with the register
  allocation result.

## Finalization and Encoding

At the end of the register allocation phase, we have enough information to
finally generate machine code (_encoding_). We are only missing the prologue
and epilogue of the function.

### Prologue and Epilogue

As usual, the **prologue** is executed before the main body of the function,

> and the **epilogue** is executed at the end. The prologue is responsible for

Reviewer: ```suggestion
and the **epilogue** is executed at the return. The prologue is responsible for
```

          |      arg 0      |               |      arg 0       | <----+
          +-----------------+               |  size_of_arg_ret |
                                            |  return address  |
                                            +------------------+ <---- SP
             (low address)                     (low address)
```

However, the prologue and epilogue might also be responsible for saving and
restoring the state of registers that might be overwritten by the function
("clobbered"); and, if spilling occurs, prologue and epilogue are also
responsible for reserving and releasing the space for the spilled values.

For clarity, we make a distinction between the space reserved for the clobbered
registers and the space reserved for the spilled values:

- Spill slots are used to temporarily store the values that needs spilling as
  determined by the register allocator. This section must have a fix height,
but its contents will change over time, as registers are being spilled and
reloaded.
- Clobbered registers are, similarly, determined by the register allocator, but
  they are stashed in the prologue and then restored in the epilogue.

> The procedure happens at the end of the register allocation phase because at
> this point we have collected enough information to know how much space we need
> to reserve.

Reviewer: ```suggestion
The procedure happens after the register allocation phase because at
this point we have collected enough information to know how much space we need
to reserve, and which registers are clobbered.
```
